File: packages/evaluatorq/src/lib/integrations/simulation/utils/prompt-builders.ts
/**
 * Prompt building utilities for personas, scenarios, and datapoints.
 *
 * These are the TypeScript equivalents of:
 * - Python Persona.to_system_prompt()
 * - Python Scenario.to_user_context()
 * - Python Datapoint.generate() / Datapoint._build_system_prompt()
 */

import type { Datapoint, Persona, Scenario } from "../types.js";
import {
  CULTURAL_CONTEXT_INSTRUCTIONS,
  EMOTIONAL_ARC_INSTRUCTIONS,
  INPUT_FORMAT_INSTRUCTIONS,
  STRATEGY_INSTRUCTIONS,
} from "../types.js";
import { delimit } from "./sanitize.js";

// ---------------------------------------------------------------------------
// Persona → system prompt
// ---------------------------------------------------------------------------

/**
 * Convert a persona to a system prompt for the user simulator.
 *
 * Mirrors Python `Persona.to_system_prompt()`.
 */
export function buildPersonaSystemPrompt(persona: Persona): string {
  const patienceDesc =
    persona.patience < 0.3
      ? "very impatient"
      : persona.patience > 0.7
        ? "patient"
        : "moderately patient";

  const assertiveDesc =
    persona.assertiveness > 0.7
      ? "very assertive and direct"
      : persona.assertiveness < 0.3
        ? "passive"
        : "balanced";

  const politeDesc =
    persona.politeness < 0.3
      ? "rude and curt"
      : persona.politeness > 0.7
        ? "very polite"
        : "neutral in tone";

  const techDesc =
    persona.technical_level < 0.3
      ? "a complete novice"
      : persona.technical_level > 0.7
        ? "a technical expert"
        : "somewhat technical";

  let arcText = "";
  const arcKey = persona.emotional_arc ?? "stable";
  const arcInstruction = EMOTIONAL_ARC_INSTRUCTIONS[arcKey];
  if (arcInstruction) {
    arcText = `\n\nEmotional Arc: ${arcInstruction}`;
  }

  let culturalText = "";
  const culturalKey = persona.cultural_context ?? "neutral";
  const culturalInstruction = CULTURAL_CONTEXT_INSTRUCTIONS[culturalKey];
  if (culturalInstruction) {
    culturalText = `\n\nCultural Communication Style: ${culturalInstruction}`;
  }

  return `You are simulating a user with the following characteristics:

Name: ${delimit(persona.name)}
Patience: You are ${patienceDesc}
Assertiveness: You are ${assertiveDesc}
Politeness: You are ${politeDesc}
Technical Level: You are ${techDesc}
Communication Style: ${delimit(persona.communication_style)}

Background: ${delimit(persona.background)}
${arcText}${culturalText}
Stay in character throughout the conversation. Your responses should reflect these traits consistently.
Do not break character or acknowledge that you are a simulation.`;
}

// ---------------------------------------------------------------------------
// Scenario → user context
// ---------------------------------------------------------------------------

const EMOTION_INSTRUCTIONS: Record<string, string> = {
  neutral: "You approach this conversation calmly.",
  frustrated: "You are frustrated and may express irritation.",
  confused: "You are confused and may ask for clarification.",
  happy: "You are in a good mood and friendly.",
  urgent: "This is urgent and you need a quick resolution.",
};

/**
 * Convert a scenario to context for the user simulator.
 *
 * Mirrors Python `Scenario.to_user_context()`.
 */
export function buildScenarioUserContext(scenario: Scenario): string {
  let criteriaText = "";
  if (scenario.criteria && scenario.criteria.length > 0) {
    const mustHappen = scenario.criteria
      .filter((c) => c.type === "must_happen")
      .map((c) => delimit(c.description));
    const mustNot = scenario.criteria
      .filter((c) => c.type === "must_not_happen")
      .map((c) => delimit(c.description));

    if (mustHappen.length > 0) {
      criteriaText += `\n\nYou expect the agent to: ${mustHappen.join(", ")}`;
    }
    if (mustNot.length > 0) {
      criteriaText += `\n\nYou would be dissatisfied if: ${mustNot.join(", ")}`;
    }
  }

  let strategyText = "";
  const strategyKey = scenario.conversation_strategy ?? "cooperative";
  const strategyInstruction = STRATEGY_INSTRUCTIONS[strategyKey];
  if (strategyInstruction) {
    strategyText = `\n\nConversation Strategy: ${strategyInstruction}`;
  }

  let formatText = "";
  const formatKey = scenario.input_format ?? "plain_text";
  const formatInstruction = INPUT_FORMAT_INSTRUCTIONS[formatKey];
  if (formatInstruction) {
    formatText = `\n\nMessage Format: ${formatInstruction}`;
  }

  const emotionText =
    EMOTION_INSTRUCTIONS[scenario.starting_emotion ?? "neutral"] ?? "";

  return `Scenario: ${delimit(scenario.name)}

Your Goal: ${delimit(scenario.goal)}

Context: ${delimit(scenario.context ?? "")}

Emotional State: ${emotionText}
${criteriaText}${strategyText}${formatText}

Work towards your goal naturally. React authentically based on how the agent responds.`;
}

// ---------------------------------------------------------------------------
// Datapoint helpers
// ---------------------------------------------------------------------------

/**
 * Build the combined system prompt from persona and scenario.
 *
 * Mirrors Python `Datapoint._build_system_prompt()`.
 */
export function buildDatapointSystemPrompt(
  persona: Persona,
  scenario: Scenario,
): string {
  const personaPrompt = buildPersonaSystemPrompt(persona);
  const scenarioContext = buildScenarioUserContext(scenario);

  return `${personaPrompt}\n\n---\n\n${scenarioContext}`;
}

/**
 * Generate a datapoint from persona and scenario.
 *
 * Mirrors Python `Datapoint.generate()`.
 */
export function generateDatapoint(
  persona: Persona,
  scenario: Scenario,
  firstMessage = "",
): Datapoint {
  return {
    id: `dp_${crypto.randomUUID().replace(/-/g, "").slice(0, 12)}`,
    persona,
    scenario,
    user_system_prompt: buildDatapointSystemPrompt(persona, scenario),
    first_message: firstMessage,
  };
}

File: packages/evaluatorq-py/src/evaluatorq/simulation/utils/prompt_builders.py
"""Prompt building utilities for personas, scenarios, and datapoints."""

from __future__ import annotations

import uuid

from evaluatorq.simulation.types import (
    CULTURAL_CONTEXT_INSTRUCTIONS,
    EMOTIONAL_ARC_INSTRUCTIONS,
    INPUT_FORMAT_INSTRUCTIONS,
    STRATEGY_INSTRUCTIONS,
    CulturalContext,
    Datapoint,
    EmotionalArc,
    InputFormat,
    Persona,
    Scenario,
    ConversationStrategy,
)
from evaluatorq.common.sanitize import delimit

# ---------------------------------------------------------------------------
# Emotion instructions (local to prompt builders)
# ---------------------------------------------------------------------------

_EMOTION_INSTRUCTIONS: dict[str, str] = {
    "neutral": "You approach this conversation calmly.",
    "frustrated": "You are frustrated and may express irritation.",
    "confused": "You are confused and may ask for clarification.",
    "happy": "You are in a good mood and friendly.",
    "urgent": "This is urgent and you need a quick resolution.",
}


# ---------------------------------------------------------------------------
# Persona → system prompt
# ---------------------------------------------------------------------------


def build_persona_system_prompt(persona: Persona) -> str:
    """Convert a persona to a system prompt for the user simulator."""
    if persona.patience < 0.3:
        patience_desc = "very impatient"
    elif persona.patience > 0.7:
        patience_desc = "patient"
    else:
        patience_desc = "moderately patient"

    if persona.assertiveness > 0.7:
        assertive_desc = "very assertive and direct"
    elif persona.assertiveness < 0.3:
        assertive_desc = "passive"
    else:
        assertive_desc = "balanced"

    if persona.politeness < 0.3:
        polite_desc = "rude and curt"
    elif persona.politeness > 0.7:
        polite_desc = "very polite"
    else:
        polite_desc = "neutral in tone"

    if persona.technical_level < 0.3:
        tech_desc = "a complete novice"
    elif persona.technical_level > 0.7:
        tech_desc = "a technical expert"
    else:
        tech_desc = "somewhat technical"

    arc_text = ""
    arc_key = persona.emotional_arc or EmotionalArc.stable
    arc_instruction = EMOTIONAL_ARC_INSTRUCTIONS.get(arc_key, "")
    if arc_instruction:
        arc_text = f"\n\nEmotional Arc: {arc_instruction}"

    cultural_text = ""
    cultural_key = persona.cultural_context or CulturalContext.neutral
    cultural_instruction = CULTURAL_CONTEXT_INSTRUCTIONS.get(cultural_key, "")
    if cultural_instruction:
        cultural_text = f"\n\nCultural Communication Style: {cultural_instruction}"

    return (
        f"You are simulating a user with the following characteristics:\n\n"
        f"Name: {delimit(persona.name)}\n"
        f"Patience: You are {patience_desc}\n"
        f"Assertiveness: You are {assertive_desc}\n"
        f"Politeness: You are {polite_desc}\n"
        f"Technical Level: You are {tech_desc}\n"
        f"Communication Style: {delimit(persona.communication_style.value)}\n\n"
        f"Background: {delimit(persona.background)}\n"
        f"{arc_text}{cultural_text}\n"
        f"Stay in character throughout the conversation. Your responses should reflect these traits consistently.\n"
        f"Do not break character or acknowledge that you are a simulation."
    )


# ---------------------------------------------------------------------------
# Scenario → user context
# ---------------------------------------------------------------------------


def build_scenario_user_context(scenario: Scenario) -> str:
    """Convert a scenario to context for the user simulator."""
    criteria_text = ""
    if scenario.criteria:
        must_happen = [
            delimit(c.description) for c in scenario.criteria if c.type == "must_happen"
        ]
        must_not = [
            delimit(c.description)
            for c in scenario.criteria
            if c.type == "must_not_happen"
        ]

        if must_happen:
            criteria_text += f"\n\nYou expect the agent to: {', '.join(must_happen)}"
        if must_not:
            criteria_text += f"\n\nYou would be dissatisfied if: {', '.join(must_not)}"

    strategy_text = ""
    strategy_key = scenario.conversation_strategy or ConversationStrategy.cooperative
    strategy_instruction = STRATEGY_INSTRUCTIONS.get(strategy_key, "")
    if strategy_instruction:
        strategy_text = f"\n\nConversation Strategy: {strategy_instruction}"

    format_text = ""
    format_key = scenario.input_format or InputFormat.plain_text
    format_instruction = INPUT_FORMAT_INSTRUCTIONS.get(format_key, "")
    if format_instruction:
        format_text = f"\n\nMessage Format: {format_instruction}"

    emotion_text = _EMOTION_INSTRUCTIONS.get(
        (scenario.starting_emotion or "neutral")
        if isinstance(scenario.starting_emotion, str)
        else (
            scenario.starting_emotion.value if scenario.starting_emotion else "neutral"
        ),
        "",
    )

    return (
        f"Scenario: {delimit(scenario.name)}\n\n"
        f"Your Goal: {delimit(scenario.goal)}\n\n"
        f"Context: {delimit(scenario.context or '')}\n\n"
        f"Emotional State: {emotion_text}\n"
        f"{criteria_text}{strategy_text}{format_text}\n\n"
        f"Work towards your goal naturally. React authentically based on how the agent responds."
    )


# ---------------------------------------------------------------------------
# Datapoint helpers
# ---------------------------------------------------------------------------


def build_datapoint_system_prompt(persona: Persona, scenario: Scenario) -> str:
    """Build the combined system prompt from persona and scenario."""
    persona_prompt = build_persona_system_prompt(persona)
    scenario_context = build_scenario_user_context(scenario)
    return f"{persona_prompt}\n\n---\n\n{scenario_context}"


def generate_datapoint(
    persona: Persona, scenario: Scenario, first_message: str = ""
) -> Datapoint:
    """Generate a datapoint from persona and scenario."""
    dp_id = f"dp_{uuid.uuid4().hex[:12]}"
    return Datapoint(
        id=dp_id,
        persona=persona,
        scenario=scenario,
        user_system_prompt=build_datapoint_system_prompt(persona, scenario),
        first_message=first_message,
    )
